### PR TITLE
mtls metrics service

### PR DESCRIPTION
**Pull request by NajiObeid**
RFD27 metrics service with optional mtls 
https://github.com/gravitational/teleport/pull/6469

### CODE REVIEW

File: lib/service/service.go [hunk 1]
```diff
@@ -1998,11 +2013,101 @@ func (process *TeleportProcess) initUploaderService(accessPoint auth.AccessPoint
...
+
+		pool := x509.NewCertPool()
+		for _, caCertPath := range process.Config.Metrics.CACerts {
+			caCert, err := ioutil.ReadFile(caCertPath)
+			if err != nil {
+				return trace.Wrap(err, "failed to read prometheus ca cert")
+			}
+
+			if !pool.AppendCertsFromPEM(caCert) {
+				return trace.Wrap(nil, "failed to parse prometheus ca cert")
```
Comment (awly) [inline lib/service/service.go:2054]:
```suggestion
				return trace.BadParameter("failed to parse prometheus ca cert %q", caCertPath)
```

File: lib/service/service.go [hunk 2]
```diff
@@ -1998,11 +2013,101 @@ func (process *TeleportProcess) initUploaderService(accessPoint auth.AccessPoint
...
+		for _, pair := range process.Config.Metrics.KeyPairs {
+			certificate, err := tls.LoadX509KeyPair(pair.Certificate, pair.PrivateKey)
+			if err != nil {
+				return trace.Wrap(err, "failed to read keypair")
+			}
+			tlsConfig.Certificates = append(tlsConfig.Certificates, certificate)
+		}
+
+		if len(tlsConfig.Certificates) == 0 {
+			return trace.Wrap(nil, "no keypairs were provided for the metrics service with mtls enabled")
```
Comment (awly) [inline lib/service/service.go:2043]:
```suggestion
			return trace.BadParameter("no keypairs were provided for the metrics service with mtls enabled")
```

File: lib/service/service.go [hunk 3]
```diff
@@ -1998,11 +2013,101 @@ func (process *TeleportProcess) initUploaderService(accessPoint auth.AccessPoint
...
+				return trace.Wrap(err, "failed to read prometheus ca cert")
+			}
+
+			if !pool.AppendCertsFromPEM(caCert) {
+				return trace.Wrap(nil, "failed to parse prometheus ca cert")
+			}
+		}
+
+		if len(pool.Subjects()) == 0 {
+			return trace.Wrap(nil, "no prometheus ca certs were provided for the metrics service with mtls enabled")
```
Comment (awly) [inline lib/service/service.go:2059]:
```suggestion
			return trace.BadParameter("no prometheus ca certs were provided for the metrics service with mtls enabled")
```

File: lib/service/service.go [hunk 4]
```diff
@@ -1998,11 +2013,101 @@ func (process *TeleportProcess) initUploaderService(accessPoint auth.AccessPoint
...
+		Handler:           mux,
+		ReadHeaderTimeout: defaults.DefaultDialTimeout,
+		ErrorLog:          utils.NewStdlogger(log.Error, teleport.ComponentMetrics),
+		TLSConfig:         tlsConfig,
+	}
+
+	log.Infof("Starting metrics service on %v.", process.Config.Metrics.ListenAddr.Addr)
+
+	process.RegisterFunc("metrics.service", func() error {
+		err := server.Serve(listener)
```
Comment (awly) [inline lib/service/service.go:2079]:
```suggestion
	}

	server := &http.Server{
		Handler:           mux,
		ReadHeaderTimeout: defaults.DefaultDialTimeout,
		ErrorLog:          utils.NewStdlogger(log.Error, teleport.ComponentMetrics),
		TLSConfig:         tlsConfig,
	}

	log.Infof("Starting metrics service on %v.", process.Config.Metrics.ListenAddr.Addr)

	process.RegisterFunc("metrics.service", func() error {
		err := server.ServeTLS(listener, "", "")
```

File: lib/config/configuration.go [hunk 5]
```diff
@@ -980,6 +987,68 @@ func applyAppsConfig(fc *FileConfig, cfg *service.Config) error {
...
+		cfg.Metrics.ListenAddr = addr
+	}
+
+	cfg.Metrics.MTLS = fc.Metrics.MTLSEnabled()
+
+	for _, p := range fc.Metrics.KeyPairs {
+		// Check that the certificate exists on disk. This exists to provide the
+		// user a sensible error message.
+		if !utils.FileExists(p.PrivateKey) {
+			return trace.Errorf("metrics service private key does not exist: %s", p.PrivateKey)
```
Comment (r0mant) [inline lib/config/configuration.go:1009]:
```suggestion
			return trace.NotFound("metrics service private key does not exist: %s", p.PrivateKey)
```

File: lib/config/configuration.go [hunk 6]
```diff
@@ -980,6 +987,68 @@ func applyAppsConfig(fc *FileConfig, cfg *service.Config) error {
...
+		}
+		if !utils.FileExists(p.Certificate) {
+			return trace.Errorf("metrics service cert does not exist: %s", p.Certificate)
```
Comment (r0mant) [inline lib/config/configuration.go:1012]:
```suggestion
			return trace.NotFound("metrics service certificate does not exist: %s", p.Certificate)
```

File: lib/config/configuration.go [hunk 7]
```diff
@@ -980,6 +987,68 @@ func applyAppsConfig(fc *FileConfig, cfg *service.Config) error {
...
+			PrivateKey:  p.PrivateKey,
+			Certificate: p.Certificate,
+		})
+	}
+
+	for _, caCert := range fc.Metrics.CACerts {
+		// Check that the certificate exists on disk. This exists to provide the
+		// user a sensible error message.
+		if !utils.FileExists(caCert) {
+			return trace.Errorf("metrics service ca cert does not exist: %s", caCert)
```
Comment (r0mant) [inline lib/config/configuration.go:1043]:
```suggestion
			return trace.NotFound("metrics service CA certificate does not exist: %s", caCert)
```

File: lib/config/fileconf.go [hunk 8]
```diff
@@ -1100,3 +1104,33 @@ func (o *OIDCConnector) Parse() (services.OIDCConnector, error) {
...
 }
+
+// Metrics is a `metrics_service` section of the config file:
+type Metrics struct {
+	// Service is a generic service configuration section
+	Service `yaml:",inline"`
+
+	// MTLSFlag states whether the metrics endpoint should be secured with mTLS
+	// or if metrics should be shipped in clear text
+	MTLSFlag string `yaml:"mtls,omitempty"`
```
Comment (r0mant) [inline lib/config/fileconf.go:1115]: Is this flag needed at all? Can we just auto-enable mTLS if key pairs and/or CAs are provided?

File: lib/config/fileconf.go [hunk 9]
```diff
@@ -1100,3 +1104,33 @@ func (o *OIDCConnector) Parse() (services.OIDCConnector, error) {
...
+	// KeyPairs is a list of x509 key pairs used for securing the metrics endpoint wihth mTLS.
+	// It should be used alongside MTLS = true
+	KeyPairs []KeyPair `yaml:"keypairs,omitempty"`
+
+	// CACerts is a list of prometheus CA certificates.
+	// It should be used alongside MTLS = true
+	CACerts []string `yaml:"ca_certs,omitempty"`
+}
+
+func (m *Metrics) MTLSEnabled() bool {
```
Comment (r0mant) [inline lib/config/fileconf.go:1126]: Nit: Add godoc on public method.

File: lib/service/cfg.go [hunk 10]
```diff
@@ -616,7 +619,6 @@ type DatabaseGCP struct {
 	InstanceID string
 }
 
-// Check validates the database proxy configuration.
```
Comment (r0mant) [inline lib/service/cfg.go:619]: Why did you remove this godoc?

File: lib/service/service.go [hunk 11]
```diff
@@ -1998,11 +2013,101 @@ func (process *TeleportProcess) initUploaderService(accessPoint auth.AccessPoint
...
+
+		if len(tlsConfig.Certificates) == 0 {
+			return trace.Wrap(nil, "no keypairs were provided for the metrics service with mtls enabled")
+		}
+
+		pool := x509.NewCertPool()
+		for _, caCertPath := range process.Config.Metrics.CACerts {
+			caCert, err := ioutil.ReadFile(caCertPath)
+			if err != nil {
+				return trace.Wrap(err, "failed to read prometheus ca cert")
```
Comment (r0mant) [inline lib/service/service.go:2050]:
```suggestion
				return trace.Wrap(err, "failed to read prometheus CA certificate: %v", caCertPath)
```

**Comment by russjones** [comment]
@NajiObeid Can you take a look and apply the requested changes? Then @awly and @r0mant can re-review.

File: lib/config/configuration.go [hunk 12]
```diff
@@ -980,6 +987,68 @@ func applyAppsConfig(fc *FileConfig, cfg *service.Config) error {
...
+			return trace.Wrap(err)
+		}
+		certificateChain, err := utils.ReadCertificateChain(certificateChainBytes)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if utils.IsSelfSigned(certificateChain) {
+			warningMessage := "Starting the metrics service with a self-signed TLS certificate."
+			log.Warnf(warningMessage)
+		} else {
```
[the remark this answers is not in the document]
Comment (NajiObeid) [reply lib/config/configuration.go:1026]: Just copied it from here
https://github.com/gravitational/teleport/blob/master/lib/config/configuration.go#L635-L639

Let me know if you want it removed

File: lib/config/fileconf.go (hunk 8 above)
Comment (NajiObeid) [reply lib/config/fileconf.go:1115]: that was the original design if I remember correctly.
it was proposed in the rfd discussion https://github.com/gravitational/teleport/pull/6469

File: lib/defaults/defaults.go [hunk 13]
```diff
@@ -63,6 +63,9 @@ const (
 	// MySQLListenPort is the default listen port for MySQL proxy.
 	MySQLListenPort = 3036
 
+	// MetricsListenPort is the default listen port for the metrics service.
+	MetricsListenPort = 3081
```
[the remark this answers is not in the document]
Comment (NajiObeid) [reply lib/defaults/defaults.go:67]: wouldn't there be a collision of both services are enabled at the same time?

[the remark this answers is not in the document]
Comment (NajiObeid) [reply lib/defaults/defaults.go:67]: I just saw your comment below about reimplementing the diag service

File: lib/service/service.go [hunk 14]
```diff
@@ -1998,11 +2013,101 @@ func (process *TeleportProcess) initUploaderService(accessPoint auth.AccessPoint
 	return nil
 }
 
+// initMetricsService starts the metrics service currently serving metrics for
+// prometheus consumption
+func (process *TeleportProcess) initMetricsService() error {
+	mux := http.NewServeMux()
+	mux.Handle("/metrics", promhttp.Handler())
```
[the remark this answers is not in the document]
Comment (NajiObeid) [reply lib/service/service.go:2020]: That was the original plan but the goal is to enable mtls on the metrics endpoint only and leave the others as is.
That can be theoretically possible if we trigger a tls renegotiation on the metrics endpoint, but it's both not supported by prometheus and also deprecated in tls 1.3

File: lib/service/service.go (hunk 4 above)
Comment (NajiObeid) [reply lib/service/service.go:2079]: I think that doesn't work if the `mtls = off` in the config. You'd be calling `server.ServerTLS` with an empty tls config and empty strings for cert  and key.

File: lib/config/configuration.go [hunk 15]
```diff
@@ -980,6 +987,68 @@ func applyAppsConfig(fc *FileConfig, cfg *service.Config) error {
...
+
+	for _, p := range fc.Metrics.KeyPairs {
+		// Check that the certificate exists on disk. This exists to provide the
+		// user a sensible error message.
+		if !utils.FileExists(p.PrivateKey) {
+			return trace.Errorf("metrics service private key does not exist: %s", p.PrivateKey)
+		}
+		if !utils.FileExists(p.Certificate) {
+			return trace.Errorf("metrics service cert does not exist: %s", p.Certificate)
+		}
```
[the remark this answers is not in the document]
Comment (NajiObeid) [reply lib/config/configuration.go:1013]: to my knowledge it doesn't. but that's how it's setup in another place in the config

File: lib/service/cfg.go (hunk 10 above)
Comment (NajiObeid) [reply lib/service/cfg.go:619]: must have been a mistake.

File: lib/service/service.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (knisbet) [reply lib/service/service.go:2013]: @awly The suggestion to use /healthz and /readyz were not part of the RFD. I don't believe it's guarenteed that a customer would want this on mTLS for healthz/readyz checks, and systems like kubernetes to my knowledge don't support mtls in healthz/readyz checks. If the option is wanted to expose those checks over mtls and not with http checks, my 2 cents is those should be individually configurable.

File: lib/config/configuration.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (awly) [reply lib/config/configuration.go:1039]: Yes, let's remove this log here.
For proxy it's different because it's a public-facing endpoint that's accessed from a browser.

File: lib/config/configuration.go [hunk 16]
```diff
@@ -992,6 +1000,78 @@ func applyAppsConfig(fc *FileConfig, cfg *service.Config) error {
...
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if utils.IsSelfSigned(certificateChain) {
+			warningMessage := "Starting the metrics service with a self-signed TLS certificate."
+			log.Warnf(warningMessage)
+		} else {
+			if err := utils.VerifyCertificateChain(certificateChain); err != nil {
+				return trace.BadParameter("unable to verify the metrics service certificate chain in %v: %s",
+					fc.Proxy.CertFile, utils.UserMessageFromError(err))
```
Comment (awly) [inline lib/config/configuration.go:1042]:
```suggestion
					p.Certificate, utils.UserMessageFromError(err))
```

File: lib/config/configuration.go [hunk 17]
```diff
@@ -992,6 +1000,78 @@ func applyAppsConfig(fc *FileConfig, cfg *service.Config) error {
...
+
+	if cfg.Metrics.MTLS {
+		if len(cfg.Metrics.KeyPairs) == 0 {
+			return trace.BadParameter("At least one keypair shoud be provided when mtls is enabled in the metrics config")
+		}
+
+		if len(cfg.Metrics.CACerts) == 0 {
+			return trace.BadParameter("At least one CA cert shoud be provided when mtls is enabled in the metrics config")
+		}
+	}
```
Comment (awly) [inline lib/config/configuration.go:1070]: nit: move this to the top, before parsing `cfg.Metrics.KeyPairs`

File: lib/config/configuration.go [hunk 18]
```diff
@@ -992,6 +1000,78 @@ func applyAppsConfig(fc *FileConfig, cfg *service.Config) error {
...
+		cfg.Metrics.CACerts = append(cfg.Metrics.CACerts, caCert)
+	}
+
+	if cfg.Metrics.MTLS {
+		if len(cfg.Metrics.KeyPairs) == 0 {
+			return trace.BadParameter("At least one keypair shoud be provided when mtls is enabled in the metrics config")
+		}
+
+		if len(cfg.Metrics.CACerts) == 0 {
+			return trace.BadParameter("At least one CA cert shoud be provided when mtls is enabled in the metrics config")
```
Comment (awly) [inline lib/config/configuration.go:1068]:
```suggestion
			return trace.BadParameter("at least one keypair shoud be provided when mtls is enabled in the metrics config")
		}

		if len(cfg.Metrics.CACerts) == 0 {
			return trace.BadParameter("at least one CA cert shoud be provided when mtls is enabled in the metrics config")
```

File: lib/service/service.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (awly) [reply lib/service/service.go:2013]: OK, I can understand leaving out `/healthz` and `/readyz`, but all the `/debug` ones seem logically closer to metrics and should be included here when both `--diag-addr` and `metrics_service` are enabled.

FWIW, I think it will be extremely confusing for on-prem customers to understand the difference between web UI, metrics and diag ports. I should've thought of this during RFD review, my bad.
Let's proceed according to the RFD, but please document in a comment and `/docs/` the difference between all these endpoints and ports.

File: lib/config/configuration.go [hunk 19]
```diff
@@ -992,6 +1000,78 @@ func applyAppsConfig(fc *FileConfig, cfg *service.Config) error {
...
+	// Metrics is enabled.
+	cfg.Metrics.Enabled = true
+
+	if fc.Metrics.ListenAddress != "" {
+		addr, err := utils.ParseHostPortAddr(fc.Metrics.ListenAddress, int(defaults.MetricsListenPort))
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		cfg.Metrics.ListenAddr = addr
+	}
```
Comment (awly) [inline lib/config/configuration.go:1014]: please remove the outer condition
`cfg.Metrics.ListenAddr` should be populated and `utils.ParseHostPortAddr` will handle defaulting when empty

**Review by awly — APPROVED** [review]

File: lib/config/fileconf.go (hunk 8 above)
Comment (r0mant) [reply lib/config/fileconf.go:1120]: Are there any downsides to enabling mtls just based on the presence of keypairs/CA? The upside is one less configuration field. @awly @russjones what do you folks think?

**Review by r0mant — COMMENTED** [review]
Just have one final UX question regarding whether we really need mtls configuration field.

File: lib/config/fileconf.go (hunk 8 above)
Comment (awly) [reply lib/config/fileconf.go:1120]: Removing the boolean field makes sense to me

**Comment by russjones** [comment]
@NajiObeid One last question, then looks like we'll be clear to merge.

File: lib/config/fileconf.go (hunk 8 above)
Comment (NajiObeid) [reply lib/config/fileconf.go:1120]: sounds good

**Comment by NajiObeid** [comment]
I think I've addressed most comments. Let me know if you think something else should be improved. I'll work on docs tomorrow. Thanks a lot everyone.

**Comment by awly** [comment]
We can add the docs as a separate PR

**Comment by NajiObeid** [comment]
Thanks @awly  let me know if all looks good on your side. 
@r0mant ?

**Review by awly — APPROVED** [review]

**Review by r0mant — APPROVED** [review]

**Comment by russjones** [comment]
@NajiObeid Go ahead and merge.
